VERSIONING.md

```diff
@@ -0,0 +1,55 @@
+Rocket Chip Versioning Strategy ** All Strawman to be discussed**
+===============================
+
+Version Naming
+--------------
+
+* Rocket Chip uses a Calendar Versioning release strategy (https://calver.org/).
+* The format is YYYY.MM.DD.MINOR
+* MINOR changes include:
+  * API Additions
+  * Bug Fixes (which can result in changes in the emitted hardware considered bugs)
+* MINOR changes do NOT include:
+  * Changes in top-level I/Os for < level of hierarchy >
+  * Backwards incompatible changes in the API (see below)
+
+API
+---
+
+The Rocket-Chip API is: 
+  * All public Scala methods and classes
+  * All public Scala methods and classes of included Scala libraries (hardfloat, CDE, etc)
+  * The command line used to invoke the generator
+  * All Verilog interfaces (I/Os and parameterization) for verilog files
```

Comment:
Is this just of the top-level module or all modules?

Reply:
There is no "top-level module" in rocket-chip which is a library of module generators. So, really I'd have to say "all modules". But yeah not sure how this could be enforced especially with diplomatic punching of things other than the previous requirement that all classes (and `Bundle`s are classes) stay the same, so maybe this is captured in that.

Actually most `val io = IO ( new Bundle {...})` are actually anonymous classes. Not sure how an anonymous class is considered in a "normal" scala API. Presumably if I added/removed a var or changed its type inside that `Bundle` that would be considered a change in the type of `val io` so would be considered a breaking change by my above rules?

In thinking about this more it's probably best to just use Scala standard definition of the API as it will probably capture *most* of the hardware use cases we think about.

Reply:
> There is no "top-level module" in rocket-chip which is a library of module generators.

Let me reword my question, since this is a bit more subtle than I had realized.

For a given "top" module (the actual module that the Chisel generation flow starts with or perhaps also any markDUT module), are the Verilog interfaces of the Verilog modules below it in that particular generation allowed to change?

For example, suppose I had the two Chisel modules (and please excuse the syntax I am making up because I don't actually know Chisel):

```scala
class OuterModule extends Module {
  val io = Bundle { val out = Output(Bool), val in1 = Input(Bool), val in2 = Input(Bool) }
  val inner_module = InnerModule()
  inner_module.io.in1 := io.in1
  inner_module.io.in2 := io.in2
  io.out := inner_module.io.out
}

class InnerModule extends Module {
  val io = Bundle { val out = Output(Bool), val in1 = Input(Bool), val in2 = Input(Bool) }
  io.out := io.in1 && io.in2
}
```

Now, suppose in a future version, we have the following change to `OuterModule`:

```diff
-  inner_module.io.in2 := io.in2
+  inner_module.io.in2 := false.asBool
```

Due to constant propagation, `InnerModule`s interface will likely change due to optimizing away the constant signal.

Now, if you instead chose `InnerModule` as your top-level module for a particular run of the generators, the IOs would _not_ change (I think?).

Reply:
I realized that this comment was in particular on a line i meant to target the verilog resources in this file (black boxes), not generated verilog.

Reply:
I am editing the text to say "verilog resources", which may not really resolve this conversation but my intention was the files which we are explicitly including as verilog here.